exec: add support for ON expression for LEFT SEMI and LEFT ANTI merge join

yuzefovich opened the pull request:
**exec: minor clean up**

**exec: add support for ON expression for LEFT SEMI and LEFT ANTI merge join**

This commit adds support for ON expression for LEFT SEMI and LEFT ANTI
merge joins. This is achieved by setting up a side chain of two operators
filterFeedOperator -> filterOperator which lives alongside the merge join
but not on the same chain. Then, whenever a filter is present, we run
the matched tuples, one at a time, from the left and right inputs through it.
When the filter expression only references the columns from the left side, we
can optimize the filtering to do the check once for every tuple in question
from the left; however, when the filter references the columns from the right,
we need to filter the full cross product of each left tuple with the full
matching group of tuples from the right which gives us quadratic behavior.

**exec: move the filtering logic out of the merge joiner**

This commit moves out the infrastructure that supports ON expressions
for merge joiner into a separate file to be reused later by hash joiner.

**distsqlrun: refactor support for ON expressions for vectorized joiners**

This commit adds a comprehensive planning step for ON expressions
for vectorized hash and merge joiners. With INNER joins, the ON
expression is supported by planning a filter after the joiner;
however, without the extra work, this can be incorrect when the
filter expression refers to the columns that are not output by the
joiner. This is now fixed. Additionally, this commit allows for
correctly populating filterOnlyOnLeft field regardless of how the
filter expression is represented.

Fixes: #39458.
Addresses: #38018.
Fixes: #40372.

Release note: None

yuzefovich committed:
exec: minor clean up

Release note: None

yuzefovich committed:
exec: add support for ON expression for LEFT SEMI and LEFT ANTI merge join

This commit adds support for ON expression for LEFT SEMI and LEFT ANTI
merge joins. This is achieved by setting up a side chain of two operators
filterFeedOperator -> filterOperator which lives alongside the merge join
but not on the same chain. Then, whenever a filter is present, we run
the matched tuples, one at a time, from the left and right inputs through it.
When the filter expression only references the columns from the left side, we
can optimize the filtering to do the check once for every tuple in question
from the left; however, when the filter references the columns from the right,
we need to filter the full cross product of each left tuple with the full
matching group of tuples from the right which gives us quadratic behavior.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/39837)
<!-- Reviewable:end -->

yuzefovich committed:
exec: move the filtering logic out of the merge joiner

This commit moves out the infrastructure that supports ON expressions
for merge joiner into a separate file to be reused later by hash joiner.

Release note: None

jordanlewis reviewed (commented):
Looks good, but seems like there are many missing tests, no? I think we will need to consider how to test this better.

> Reviewed 3 of 3 files at r1, 1 of 8 files at r3, 7 of 7 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39837#-:-Ln3_Itt2cO1xOijVCp9:b-ywd3wl)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @solongordon, and @yuzefovich)

___
*[pkg/sql/exec/mergejoiner.go, line 173 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/39837#-Ln3WURO6RReHYOQ_t4O:-Ln3WURO6RReHYOQ_t4P:b-qv6tup) ([raw file](https://github.com/cockroachdb/cockroach/blob/9591d932bc58fcac37cbbc0fa2204df0a646280d/pkg/sql/exec/mergejoiner.go#L173)):*
> ```Go
> // filterFeedOperator is used to feed the filter by manually setting the batch
> // to be returned on the first Next() call with zero-length batches returned on
> // all consequent calls.
> ```

s/consequent/subsequent/
___
*[pkg/sql/exec/mergejoiner.go, line 423 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/39837#-Ln3Wtc8E6ud8OYiiRf_:-Ln3Wtc8E6ud8OYiiRfa:b-s47m5) ([raw file](https://github.com/cockroachdb/cockroach/blob/9591d932bc58fcac37cbbc0fa2204df0a646280d/pkg/sql/exec/mergejoiner.go#L423)):*
> ```Go
> 	// input uses only two batches of lengths 1 and 0 and the filter itself does
> 	// not use any static memory.
> 	// TODO(yuzefovich): update this calculation if the above comment is no
> ```

The filter could use arbitrary static memory since it can have arbitrary expressions inside. I think you need to add the filter's static memory too.
___
*[pkg/sql/exec/mergejoiner_tmpl.go, line 1043 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39837#-LmzWIgE9YRbP0a18SpC:-LmzWIgE9YRbP0a18SpD:b-3lifo5) ([raw file](https://github.com/cockroachdb/cockroach/blob/64cd2a1a84aab0051687e4d945e26a097e017a0e/pkg/sql/exec/mergejoiner_tmpl.go#L1043)):*
> ```Go
> // exhausted.
> func (o *mergeJoin_JOIN_TYPE_STRINGOp) exhaustLeftSource() {
> 	// {{ if or _JOIN_TYPE.IsInner _JOIN_TYPE.IsLeftSemi }}
> ```

How come we didn't notice this before? Are we missing tests?
___
*[pkg/sql/exec/mergejoiner_tmpl.go, line 1112 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/39837#-Ln3Zs5E4GyJEwqVflOW:-Ln3Zs5E4GyJEwqVflOX:bhgbr09) ([raw file](https://github.com/cockroachdb/cockroach/blob/9591d932bc58fcac37cbbc0fa2204df0a646280d/pkg/sql/exec/mergejoiner_tmpl.go#L1112)):*
> ```Go
> 	for lIdx := 0; lIdx < int(o.proberState.lBufferedGroup.length); lIdx++ {
> 		if o.isLeftTupleFilteredOut(ctx, o.proberState.lBufferedGroup, rBatch, lIdx, 0 /* rStartIdx */, rEndIdx) {
> 			o.builderState.lGroups = append(o.builderState.lGroups, group{
> ```

I don't think I understand why we have to filter both in `setBuilderSourceToBufferedGroup` and in `probe_switch` - maybe this is when we can pre-filter without knowing the right side? but in that case I'm confused why we still do this even if `filterOnlyOnLeft` is false.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
I agree about the tests, but I don't have any bright ideas yet on better test coverage.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39837#-:-Ln3fY_Q1Jw-bS5KjtGI:b-ree2)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis and @solongordon)

___
*[pkg/sql/exec/mergejoiner.go, line 173 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/39837#-Ln3WURO6RReHYOQ_t4O:-Ln3bCv33X-Lkni7fAmv:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9591d932bc58fcac37cbbc0fa2204df0a646280d/pkg/sql/exec/mergejoiner.go#L173)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

s/consequent/subsequent/
</blockquote></details>

Done.
___
*[pkg/sql/exec/mergejoiner.go, line 423 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/39837#-Ln3Wtc8E6ud8OYiiRf_:-Ln3blRc7cJ7PRY-NWyE:bfbq99i) ([raw file](https://github.com/cockroachdb/cockroach/blob/9591d932bc58fcac37cbbc0fa2204df0a646280d/pkg/sql/exec/mergejoiner.go#L423)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

The filter could use arbitrary static memory since it can have arbitrary expressions inside. I think you need to add the filter's static memory too.
</blockquote></details>

Good point, done.
___
*[pkg/sql/exec/mergejoiner_tmpl.go, line 1043 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/39837#-LmzWIgE9YRbP0a18SpC:-Ln3a7DE8sW3krT1tW0l:b-hznv9l) ([raw file](https://github.com/cockroachdb/cockroach/blob/64cd2a1a84aab0051687e4d945e26a097e017a0e/pkg/sql/exec/mergejoiner_tmpl.go#L1043)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

How come we didn't notice this before? Are we missing tests?
</blockquote></details>

This is a stylistic change - previously, no code was generated in case of LEFT SEMI, and it is still the case here. Now it's just explicitly pointed out in the comment.
___
*[pkg/sql/exec/mergejoiner_tmpl.go, line 1112 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/39837#-Ln3Zs5E4GyJEwqVflOW:-Ln3cE_-7RSy-YtZHzoq:bm8eavp) ([raw file](https://github.com/cockroachdb/cockroach/blob/9591d932bc58fcac37cbbc0fa2204df0a646280d/pkg/sql/exec/mergejoiner_tmpl.go#L1112)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

I don't think I understand why we have to filter both in `setBuilderSourceToBufferedGroup` and in `probe_switch` - maybe this is when we can pre-filter without knowing the right side? but in that case I'm confused why we still do this even if `filterOnlyOnLeft` is false.
</blockquote></details>

The buffering step (i.e. `appendToBufferedGroup`) occurs before we check the filter in `_PROBE_SWITCH`. In general case, I think we first need to fully buffer the groups from both sides, and only after that we can run the filter to see which tuples should be kept. Consider a case when we have a filter which checks that a column from the left equals to a column on the right (at least one not an equality column) and set up is such that the match occurs on the second batch from the right. After we fully buffered up the first batches, we don't have enough information to decide whether the tuple from the left is filtered out. I guess we could run a filter with "one right batch at a time", but I'm not sure it'll be faster or more comprehensible.

However, when `filterOnlyOnLeft` is true, we can do the filtering before appending to the buffered group. But, again, I'm not sure whether it'll be more efficient or understandable.


<!-- Sent from Reviewable.io -->

yuzefovich commented:
I'm running some queries having only LEFT SEMI join with ON expression on TPCH dataset, and the vectorized is actually less performant than the row engine :/

jordanlewis commented:
That's very surprising - take some cpu profiles and see what's up? Maybe it's because of the row-at-a-time filter stuff, or we're doing something else wrong. Feel free to upload the cpu profile and I can also take a look.

yuzefovich commented:
Actually, to me it's not that surprising - I feared that the current approach is too inefficient.

I looked into the profile, and the slowdown is because, currently, for every tuple to check, we're slicing out a batch. 
<img width="564" alt="Screen Shot 2019-08-24 at 1 15 25 PM" src="https://user-images.githubusercontent.com/5135385/63642242-443d8f80-c671-11e9-954d-4c481f906b12.png">

I probably need to write a "windowing" operator to be an input to the filter, and that operator would provide a window of a single tuple into the batch. However, I do feel that it can wait, and we could go ahead with the current approach and work on the performance later. At the moment, I'm more concerned with the correctness issues.

yuzefovich commented:
I've tried several queries with LEFT SEMI and LEFT ANTI merge joins with ON expressions on the TPCH dataset, and they seem to be returning the correct results, so I'm relatively confident in correctness of this PR. I also saw LEFT ANTI hash join query return different results for non-vec vs vec, so I'm guessing that query 21 returned different results because of the hash joiner. (As a side note, flat bytes introduced some changes that break the merge joiner).

yuzefovich commented:
Friendly ping. Let's merge this as is, and I will address the size of the generated files next.

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39837)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis and @solongordon)

___
*[pkg/sql/distsqlrun/column_exec_setup.go, line 421 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/39837#-LnUFKVg7S_tA82eSTXv:-LnUFKVg7S_tA82eSTXw:b-hbz8a7) ([raw file](https://github.com/cockroachdb/cockroach/blob/a6b1046169813b3c104248744ace944d7a4ac307/pkg/sql/distsqlrun/column_exec_setup.go#L421)):*
> ```Go
> 			// TODO(yuzefovich): figure out how to correctly populate
> 			// filterOnlyOnLeft when core.MergeJoiner.OnExpr.LocalExpr is non-nil.
> 			onExpr := core.MergeJoiner.OnExpr.Expr
> ```

Addressing the TODO above is actually important for correctness. When `filterOnlyOnLeft` is set to true, the filter input will only contain the columns from the left input, so if we have a situation when `OnExpr.Expr` is nil and `OnExpr.LocalExpr` is non-nil, but the expression refers to the columns from the right, then (right now) we set `filterOnlyOnLeft` to true, and the filter might either return incorrect results or an error.


<!-- Sent from Reviewable.io -->

jordanlewis commented:
> Addressing the TODO above is actually important for correctness. 

Does this mean this cannot merge until it is resolved? I don't quite understand.

yuzefovich reviewed (commented):
Yes, I believe so. How would I go about addressing that TODO? I would need to write a visitor that would walk over the `TypedExpr`?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39837#-:-LnUK_ykCCBwcdsu5aBz:b-sbonnt)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis and @solongordon)



<!-- Sent from Reviewable.io -->

jordanlewis commented:
Ah, I see... perhaps we wish to disable the `LocalExpr` optimization for vectorized instead as a simplification.

yuzefovich commented:
I think for now it is easier to just reject running such queries through the vectorized.

I should be good to do. PTAL.

yuzefovich commented:
Friendly ping.

jordanlewis commented:
Sorry for the delay.

yuzefovich committed:
distsqlrun: refactor support for ON expressions for vectorized joiners

This commit adds a comprehensive planning step for ON expressions
for vectorized hash and merge joiners. With INNER joins, the ON
expression is supported by planning a filter after the joiner;
however, without the extra work, this can be incorrect when the
filter expression refers to the columns that are not output by the
joiner. This is now fixed. Additionally, this commit allows for
correctly populating filterOnlyOnLeft field regardless of how the
filter expression is represented.

Release note: None

yuzefovich commented:
I added a commit to fix the planning of ON expressions to this PR because the work allows us to populate `filterOnlyOnLeft` field in all cases easily. Let me know if you think it should be a separate PR.

jordanlewis reviewed (commented):
Generally LGTM, but I wonder, is this a risky change? Also, I have a question about "internal columns" and whether we need to be as tricky as we are being here.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39837#-:-LnzBDsoEGI9Lld7K1ke:bv493cq)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @solongordon, and @yuzefovich)

___
*[pkg/sql/distsqlrun/column_exec_setup.go, line 727 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/39837#-LnzXZjQ3eNcgz4VfZZd:-LnzXZjR8-GdXS9f76r4:b-j8sz1d) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c8c4d480d66e903692449fcfe76b4a78f3b7a97/pkg/sql/distsqlrun/column_exec_setup.go#L727)):*
> ```Go
> // out after the filter has been run.
> // It returns updated leftOutCols and rightOutCols.
> func (p *filterPlanningState) renderAllNeededCols(
> ```

This seems quite clever. I wonder if we're being too clever. The MergeJoinerSpec claims that the "internal columns" of the processor are *all* of the left and right side columns. Would it be easier and simpler to just output all of them? Or do you think it would be too expensive?


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
Hm, I don't think of this PR as being too risky because I ran `TestMergeJoinerAgainstProcessor` for a couple of thousand of iterations with ON expressions with no failures (granted, there is always a single ON expression), so I'm quite confident in the correctness. Also, there shouldn't be any stability concerns because we're catching all the panics, and if an internal error occurs, then, oh well :man_shrugging: 

Technically speaking, we're still in "feature development" phase, so it should be OK for this to go in, I think :)

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39837#-:-Lo1WpaT7JHJBVK3EbCD:b-jzdhw8)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis and @solongordon)

___
*[pkg/sql/distsqlrun/column_exec_setup.go, line 727 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/39837#-LnzXZjQ3eNcgz4VfZZd:-Lo1VK_VDvAa2ZMhp4UV:bef7dbc) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c8c4d480d66e903692449fcfe76b4a78f3b7a97/pkg/sql/distsqlrun/column_exec_setup.go#L727)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

This seems quite clever. I wonder if we're being too clever. The MergeJoinerSpec claims that the "internal columns" of the processor are *all* of the left and right side columns. Would it be easier and simpler to just output all of them? Or do you think it would be too expensive?
</blockquote></details>

It would definitely be simpler to output all of the columns, and that's what I initially thought of doing, but I do believe this will be too expensive.

I guess if we want to be on the safer side right now, we could output all the columns for 19.2 release and then use this trickier way for 20.1 so that we have more confidence in it, but I'm really concerned about the performance impact of such change (I feel like we could decrease the performance of the merge join by something like 2-4x in some cases, just a non-scientific gut feeling).


<!-- Sent from Reviewable.io -->

jordanlewis reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/39837#-:-Lo62ruiEdO_cR-oi9ET:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jordanlewis and @solongordon)



<!-- Sent from Reviewable.io -->

yuzefovich commented:
Thanks for the review!

bors r+

yuzefovich commented:
bors r-

yuzefovich commented:
I needed to update the error message in `tpch_vec` logic test for query 21 now that it gets a little further in the planning.

bors r+

yuzefovich commented:
bors seems to have been stuck.

bors r-

yuzefovich commented:
bors r+
